### DISCUSSION TITLE
Using a dependency manager for C++

### DISCUSSION THREAD

**Discussion by arnaudvergnet** [organicmaps/organicmaps#2394] [Ideas]
I come from a Python/Java/JavaScript background and all those languages can handle dependencies very well (pip, gradle/maven, npm/yarn).

Looking at the project I see C++ dependencies are simply copy-pasted or using git submodules. It seems to be working but it adds so much overhead to the repo it makes new contributions more difficult. The 3party folder is about 1GB and the .git folder is about 4GB.

Why not use a C++ package manager such as [Buckaroo](https://buckaroo.pm/) or [Conan](https://conan.io/)? This should allow to clean up the repo and make non-code contributions such as translations easier. It would also reduce the effort of upgrading existing dependencies.

But then again I have very few experience with C++ dependency management so I would love to hear your reasoning.

**Comment by biodranik** [comment]
There are several reasons:
1. There were no package managers for C++ when we started the project.
2. Existing package managers have some issues. Generally, it's a very complex task to make a good cross-platform package manager for C/C++.
3. Not every dependent library is available in package managers.
4. We often patch 3party libs for some specific cases or features. These patches are not always available upstream. So we can't use an unpatched version from the package manager.

The .git folder is large because the project has a long history, and we store binary world maps with each new release in git historically.

What is wrong or what worries you (except disk space) in the current approach? Cloning repo with submodules and running a configure script is not a big deal.

**Reply by arnaudvergnet** [reply]
1. Yeah even now there is no dominant package manager
2. From what I could see from a quick search they seem to work cross-platform
3. Seems to be possible to use a github repo as source ([here is an example for bukaroo](https://github.com/LoopPerfect/buckaroo/wiki/Commands#add))
4. Patching external library does not seem like a good idea. Why not send the patches upstream?

Most the languages I tried used a package manager to handle dependencies so not having one feels like it would be difficult to manage everything. I am especially thinking about new contributors: the easier it is to understand and clone the repo, the more they will start contributing and using a package manager helps a lot.

**Reply by biodranik** [reply]
You may need patches for different reasons. For example, we heavily patched a Twine tool to properly support our cases. Or recently applied a small patch for pugixml library to support std::string_view. Publish patches into the upstream is a good idea but it should not block the development of the product. Right?

I do not understand the issue with new contributors. It's not a NodeJS. We rarely add 3party libs, and carefully choose each one. In 99% of cases, contributors don't even need to worry about 3party libs. Can you please provide a more concrete example?

**Reply by arnaudvergnet** [reply]
I was mainly thinking about someone wanting to improve translation. In this case they will need to download all 3party cpp libraries. This makes the cloning process a lot longer.

About patches I agree that patching for immediate support can be nice I just have not encountered this case before.

**Reply by biodranik** [reply]
Improving translations is easy if only necessary text files are edited. It could be even easier with some online tools like Weblate (but there are some issues with them ATM).

**Reply by arnaudvergnet** [reply]
Thank you for your answers. I just experienced an issue with submodules which took me way too much time to understand. A week ago you switched pugxml to a submodule. But doing a `git pull --recurse-submodules` would not register the new submodule so I was left with compilation errors for missing files.

In this case I had to manually run `git submodule update --recursive`. For people like me not used to working with submodules this should be either documented or automated with an update script. Another issue a package manager would solve ;)

**Reply by biodranik** [reply]
It is git basics knowledge :) If you use some tool, you need to know it. The same will be true for any package manager.

Please don't mind asking if you don't understand something to save time next time.
